gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GCSJavaClientImpl.java

```diff
@@ -0,0 +1,223 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.auth.Credentials;
+import com.google.cloud.hadoop.util.AccessBoundary;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.flogger.GoogleLogger;
+import java.io.IOException;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.channels.WritableByteChannel;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Provides read/write access to Google Cloud Storage (GCS), using Java nio channel semantics. This
+ * is a basic implementation of the GoogleCloudStorage interface that mostly delegates through to
+ * the appropriate API call(s) google-cloud-storage client.
+ */
+@VisibleForTesting
+public class GCSJavaClientImpl implements GoogleCloudStorage {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  /**
+   * Having an instance of gscImpl to redirect calls to Json client while new client implementation
+   * is in WIP.
+   */
+  private GoogleCloudStorageImpl gcsClientDelegate;
+
+  private GoogleCloudStorageOptions storageOptions;
+  private Credentials credentials;
+
+  private GCSJavaClientImpl(GCSJavaClientImplBuilder builder) throws IOException {
+    this.storageOptions = checkNotNull(builder.storageOptions, "options must not be null");
+    this.credentials = checkNotNull(builder.credentials, "credentials must not be null");
```

Comment:
if httpRequestInitializer is not null, this can be null?

Reply:
Didn't understand it fully.
`httpRequestInitializer` will and should be null for all production usecases, We exposed a constructor in GCSImpl for test use cases only.

Reply:
if builder. httpRequestInitializer is not null, the credentials are not used. So I was wondering why one need to provide credentials if httpRequestInitializer is provided.

Reply:
`Short ans`: because we don't want to have unnecessary checks in code for just testing usecases.
`Long ans`: If we are overriding `httpRequestInitializer` we must have used credentials to create it (there is no workaround to that). Now, we can handle this requirement in two ways
1. Create constructor specifically for this testing usecase. That is what is done in GCSImpl.
2. Second is, don't introduce extra code in production and instead try to align testing as per production code. That's what I am trying to achieve here by imposing testing code to pass in credentials too which they any how would have if they are overriding `httpRequestInitializer`.